flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/strategy/AdaptiveBroadcastJoinOptimizationStrategy.java

```diff
@@ -0,0 +1,213 @@
+                            .get(adaptiveJoinNode.getId())
+                            .get(1);
+            Long rightInputSize =
+                    aggregatedProducedBytesByTypeNumberAndNodeId
+                            .get(adaptiveJoinNode.getId())
+                            .get(2);
+            Preconditions.checkArgument(
+                    leftInputSize != null && rightInputSize != null,
+                    "Adaptive join node currently supports only two inputs, "
+                            + "but received input bytes with left [%s] and right [%s] for stream "
+                            + "node id [%s].",
+                    leftInputSize,
+                    rightInputSize,
+                    adaptiveJoinNode.getId());
+
+            boolean leftSizeSmallerThanThreshold = leftInputSize <= broadcastThreshold;
+            boolean rightSizeSmallerThanThreshold = rightInputSize <= broadcastThreshold;
+            boolean leftSmallerThanRight = leftInputSize < rightInputSize;
+            FlinkJoinType joinType = adaptiveJoin.getJoinType();
+            boolean canBeBroadcast;
+            boolean leftIsBuild;
+            switch (joinType) {
+                case RIGHT:
+                    // For a right outer join, if the left side can be broadcast, then the left side
+                    // is
+                    // always the build side; otherwise, the smaller side is the build side.
+                    canBeBroadcast = leftSizeSmallerThanThreshold;
+                    leftIsBuild = true;
+                    break;
+                case INNER:
+                    canBeBroadcast = leftSizeSmallerThanThreshold || rightSizeSmallerThanThreshold;
+                    leftIsBuild = leftSmallerThanRight;
+                    break;
+                case LEFT:
+                case SEMI:
+                case ANTI:
+                    // For left outer / semi / anti join, if the right side can be broadcast, then
+                    // the
+                    // right side is always the build side; otherwise, the smaller side is the build
+                    // side.
+                    canBeBroadcast = rightSizeSmallerThanThreshold;
+                    leftIsBuild = false;
+                    break;
+                case FULL:
+                default:
+                    throw new RuntimeException(String.format("Unexpected join type %s.", joinType));
+            }
+
+            boolean isBroadcast = false;
+            if (canBeBroadcast) {
+                isBroadcast =
+                        tryModifyStreamEdgesForBroadcastJoin(
+                                adaptiveJoinNode.getInEdges(), context, leftIsBuild);
+
+                if (isBroadcast) {
+                    LOG.info(
+                            "The {} input data size of the join node [{}] is small enough which is {} bytes, "
+                                    + "adaptively convert it to a broadcast hash join.",
```

Comment:
Maybe 'The {} input data size of the join node [{}] is small enough, adaptively convert it to a broadcast hash join. Broadcast threshold bytes: {}, left input bytes: {}, right input bytes: {}.'